core/camel-xml-io/src/main/java/org/apache/camel/xml/LwModelToXMLDumper.java

```diff
@@ -258,4 +283,55 @@ private static NamespaceAware getNamespaceAwareFromExpression(ExpressionNode exp
         return na;
     }
 
+    private static class BeanModelWriter implements CamelContextAware {
+
+        private final StringWriter buffer;
+        private CamelContext camelContext;
+
+        public BeanModelWriter(StringWriter buffer) {
+            this.buffer = buffer;
+        }
+
+        @Override
+        public CamelContext getCamelContext() {
+            return camelContext;
+        }
+
+        @Override
+        public void setCamelContext(CamelContext camelContext) {
+            this.camelContext = camelContext;
+        }
+
+        public void start() {
+            // noop
+        }
+
+        public void stop() {
+            // noop
+        }
+
+        public void writeBeans(List<RegistryBeanDefinition> beans) {
+            if (beans.isEmpty()) {
+                return;
+            }
+            for (RegistryBeanDefinition b : beans) {
+                doWriteRegistryBeanDefinition(b);
+            }
+        }
+
+        private void doWriteRegistryBeanDefinition(RegistryBeanDefinition b) {
+            buffer.write(String.format("    <bean name=\"%s\" type=\"%s\">%n", b.getName(), b.getType()));
+            if (b.getProperties() != null && !b.getProperties().isEmpty()) {
+                buffer.write(String.format("        <properties>%n"));
+                for (Map.Entry<String, Object> entry : b.getProperties().entrySet()) {
+                    String key = entry.getKey();
+                    Object value = entry.getValue();
+                    buffer.write(String.format("            <property key=\"%s\" value=\"%s\"/>%n", key, value));
+                }
+                buffer.write(String.format("        </properties>%n"));
+            }
+            buffer.write(String.format("    </bean>%n"));
+        }
+    }
```

Comment:
@davsclaus this is a little bit off-topic but, I think it's something we should start thinking about for the future (maybe we'll need to create a ticket about it). 

We have a lot of places with code-to-be-generated that are built-in in the code base. IMHO, this is extremely hard to read, maintain and debug. At some point in 4.x, I believe, we should start investigating how to move this to some kind of template library.

Reply:
I created a ticket to take note and simplify the discussion: https://issues.apache.org/jira/browse/CAMEL-19779